aten/src/ATen/native/cuda/TensorTopK.cu

```diff
@@ -162,10 +171,325 @@ __global__ void gatherTopK(at::cuda::detail::TensorInfo<T, IndexType> input,
+      if (current_bit > 0) {
+        desires[slice_idx] = desired;
+        ks_to_find[slice_idx] = k_to_find - digit_count_cumsum_left;
+      } else {
+        kthValues[slice_idx] = TopKTypeConfig<T>::deconvert(desired);
+      }
+    }
+  }
+
+  // reset semaphores for the next pass
+  if (tidx == 0) {
+    semaphores[slice_idx] = 0;
+  }
+};
+
+template <typename T, typename IndexType, int Dim>
+void launch(
+    at::cuda::detail::TensorInfo<T, IndexType> input,
+    IndexType inputSliceSize,
+    IndexType outputSliceSize, // aka `k`
+    bool largest,
+
+    IndexType numInputSlices,
+    IndexType inputWithinSliceStride,
+
+    at::cuda::detail::TensorInfo<T, IndexType> topK,
+    IndexType topKWithinSliceStride,
+
+    at::cuda::detail::TensorInfo<int64_t, IndexType> indices,
+    IndexType indicesWithinSliceStride) {
+
+  // configure items_per_thread based on device architecture and input size
+  // occupancy of this kernel is limited by registers per threads
+  constexpr int REGS_PER_THREAD = 40; // from nsight launch statistics
+  constexpr int REGS_PER_BLOCK = REGS_PER_THREAD * BLOCK_THREADS;
+  cudaDeviceProp* prop = at::cuda::getCurrentDeviceProperties();
+  int mpc = prop->multiProcessorCount;
+#if defined(USE_ROCM)
+  int regs_per_mp = prop->regsPerBlock;
+  int blocks_per_mp = regs_per_mp / REGS_PER_BLOCK;
+#else
+  int regs_per_mp = prop->regsPerMultiprocessor;
+  int blocks_per_mp = std::min(regs_per_mp / REGS_PER_BLOCK, prop->maxBlocksPerMultiProcessor);
+#endif
+  int64_t items_per_thread = at::ceil_div((int64_t)(inputSliceSize * numInputSlices), (int64_t)(mpc * blocks_per_mp * BLOCK_THREADS));
+  items_per_thread = std::max(4, std::min((int)items_per_thread, 64)); // clamp to (4, 64)
+  int items_per_block = items_per_thread * BLOCK_THREADS;
+
+  using Bitwise = typename TopKTypeConfig<T>::RadixType;
+  int64_t blocks_per_slice = at::ceil_div((int64_t)inputSliceSize, (int64_t)items_per_block);
+  int64_t num_blocks = numInputSlices * blocks_per_slice;
+
+  // temporary storage
+  auto& allocator = *c10::cuda::CUDACachingAllocator::get();
+
+  auto kthValues_buffer = allocator.allocate(numInputSlices * sizeof(T));
+  T* kthValues = reinterpret_cast<T*>(kthValues_buffer.get());
+
+  TORCH_CHECK(blocks_per_slice <= std::numeric_limits<uint32_t>::max(), "blocks_per_slice larger than uint32 maximum is not supported");
+  auto semaphores_buffer = allocator.allocate(numInputSlices * sizeof(uint32_t));
+  uint32_t* semaphores = reinterpret_cast<uint32_t*>(semaphores_buffer.get());
+  AT_CUDA_CHECK(cudaMemsetAsync(semaphores, 0, numInputSlices * sizeof(uint32_t), c10::cuda::getCurrentCUDAStream()));
+
+  auto ks_to_find_buffer = allocator.allocate(numInputSlices * sizeof(IndexType));
+  IndexType* ks_to_find = reinterpret_cast<IndexType*>(ks_to_find_buffer.get());
+  IndexType k_to_find = largest ? inputSliceSize - outputSliceSize + 1: outputSliceSize;
```

Comment:
The performance should be similar for largest and smllest key. Because in the kernel, finally it is only comparing the prefix_sum for `2^RADIX_BITS=256` numbers with 256 threads. so the order should not matter that much.
https://github.com/pytorch/pytorch/pull/71081/files#diff-28b350b5a7085a67679fe1932e1d60a924a1724fef619ed980d0a512af3d8189R358

the benchmarks are using `largest=True`